addon/components/ember-ace.js

```diff
@@ -92,8 +92,25 @@ export default Component.extend({
 
     this._syncAceProperties();
 
+    let settingValue = false;
+    let newValue = null;
+
+    const originalSetValue = editor.setValue;
+    editor.setValue = function(value) {
+      newValue = value;
+      settingValue = true;
+      originalSetValue.call(editor, ...arguments);
+      settingValue = false;
+    }
+
     editor.getSession().on('change', (event, session) => {
       const update = this.get('update');
+
+      // Ace implements document.setValue by first removing and then inserting
+      // These checks prevent ember-ace from triggering 'update' more than once per setValue
+      if (settingValue && newValue && event.action === 'remove') {
```

Comment:
What is the `newValue` check here doing? I don't think I fully understand the flow here, so feel free to tell me to get lost, but it seems like this could maybe join with the logic below to become something like

```js
editor.setValue = (value) => {
  const wasSilenced = this._silenceUpdates;
  this._silenceUpdates = true;
  originalSetValue.apply(this, arguments);
  this._silenceUpdates = wasSilenced;
};

editor.getSession().on('change', (event, session) => {
  const update = this.get('update');

  if (update && !this._silenceUpdates && event.action !== 'remove') {
    run(...);
  }
});
```

Reply:
Calling `editor.setValue` creates two 'change' events. One is a remove and one is an insert.

We want to silence the first update event (the 'remove' with a truthy value), but let the second update event (the 'insert' event with a truthy value) through.

As far as I can tell, there's no way to determine in setValue if we want to completely silence updates or not since we only want to silence one of the 'update' events that it triggers.


Calling `editor.setValue` creates one 'change' event, which is a remove. Here, setValue is called with a falsy value and we want to let that through as usual.

Reply:
I see, so `setValue('')` results in different behavior from Ace than `setValue(<anything-truthy>)` 🤔 

Could we _always_ silence updates that `setValue` would normally trigger and then explicitly call the `update` action afterwards ourselves?

Reply:
@dfreeman PR updated